vowpalwabbit/ftrl.cc

```diff
@@ -92,18 +94,22 @@ void multipredict(
     const auto& simple_red_features = ec._reduction_features.template get<simple_label_reduction_features>();
     pred[c].scalar = simple_red_features.initial;
   }
+  size_t num_features_from_interactions = 0;
   if (b.all->weights.sparse)
   {
     GD::multipredict_info<sparse_parameters> mp = {
         count, step, pred, all.weights.sparse_weights, static_cast<float>(all.sd->gravity)};
-    GD::foreach_feature<GD::multipredict_info<sparse_parameters>, uint64_t, GD::vec_add_multipredict>(all, ec, mp);
+    GD::foreach_feature<GD::multipredict_info<sparse_parameters>, uint64_t, GD::vec_add_multipredict>(
+        all, ec, mp, num_features_from_interactions);
   }
   else
   {
     GD::multipredict_info<dense_parameters> mp = {
         count, step, pred, all.weights.dense_weights, static_cast<float>(all.sd->gravity)};
-    GD::foreach_feature<GD::multipredict_info<dense_parameters>, uint64_t, GD::vec_add_multipredict>(all, ec, mp);
+    GD::foreach_feature<GD::multipredict_info<dense_parameters>, uint64_t, GD::vec_add_multipredict>(
+        all, ec, mp, num_features_from_interactions);
   }
+  ec.num_features_from_interactions = num_features_from_interactions;
```

Comment:
Valid concern, I thought the same. But I think there are situations where foreach_feature is called and you don't want to set the num_feature_from_interactions. Additionally it prevents examples from being const where they otherwise can  be

Reply:
do you think it would be worth to wrap those calls?

(a) where examples can be const there is a non-const overload that calls the const version and then sets the number of features on the example and
(b) if a foreach call doesn't want to set the number of features we could again overload and just not set the example with the result

Reply:
The feature count is a very non-critical operation and it's only used as a measure of work. If it's missed really all that happens is the progressive validation log will show fewer features. There are already quite a few overloads of foreach_feature and this PR is adding more overloads there. I don't think it would be worth it